temporal-sdk/src/test/java/io/temporal/workflow/TestWorkflowRule.java

```diff
@@ -0,0 +1,169 @@
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.workflow;
+
+import io.temporal.client.WorkflowClient;
+import io.temporal.client.WorkflowClientOptions;
+import io.temporal.testing.TestEnvironmentOptions;
+import io.temporal.testing.TestWorkflowEnvironment;
+import io.temporal.worker.Worker;
+import io.temporal.worker.WorkerFactoryOptions;
+import io.temporal.worker.WorkerOptions;
+import java.util.UUID;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public class TestWorkflowRule implements TestRule {
+
+  private final Class<?>[] workflowTypes;
+  private final Object[] activityImplementations;
+  private final TestWorkflowEnvironment testEnvironment;
+  private final WorkerOptions workerOptions;
+  private final boolean useExternalService;
+  private Worker worker;
+  private String taskQueue;
+
+  private TestWorkflowRule(
+      TestWorkflowEnvironment testEnvironment,
+      boolean useExternalService,
+      Class<?>[] workflowTypes,
+      Object[] activityImplementations,
+      WorkerOptions workerOptions) {
+    this.testEnvironment = testEnvironment;
+    this.useExternalService = useExternalService;
+    this.workflowTypes = workflowTypes;
+    this.activityImplementations = activityImplementations;
+    this.workerOptions = workerOptions;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private WorkerOptions workerOptions;
+    private String namespace;
+    private Class<?>[] workflowTypes;
+    private Object[] activityImplementations;
+    private boolean useExternalService;
+    private String target;
+
+    private Builder() {}
+
+    public Builder setWorkerOptions(WorkerOptions options) {
+      this.workerOptions = options;
+      return this;
+    }
+
+    public Builder setNamespace(String namespace) {
+      this.namespace = namespace;
+      return this;
+    }
+
+    public Builder setWorkflowTypes(Class<?>... workflowTypes) {
+      this.workflowTypes = workflowTypes;
+      return this;
+    }
+
+    public Builder setActivityImplementations(Object... activityImplementations) {
+      this.activityImplementations = activityImplementations;
+      return this;
+    }
+
+    public Builder setUseExternalService(boolean useExternalService) {
+      this.useExternalService = useExternalService;
+      return this;
+    }
+
+    public Builder setTarget(String target) {
```

Comment:
Add JavaDoc explaining the meaning of the target and useExternalService properties.

Reply:
Added here as well  as also included a java doc with example for the entire rule.